### ISSUE TITLE
TPM2.0 usage

### ISSUE THREAD

**Issue by atomsoftwarestudios** [01org/tpm2.0-tools#56]
Hi

I have a little problem with TPM usage, hopefully somebody can help me.

a) When I issue the tpm2_listpcrs, all pcrs are set to zero (except 17-22 which are set to FF FF...)
   I would expect that at least few PCRs would be set by BIOS during the boot process. Am I right or do I understand the concept in a wrong way?

b) can somebody help me to setup the policy for reading of the NV Index? I would like to allow access to the data just in case the PCRs will be valid (will match the values previously measured during the boot process). Can I achieve this with TPM2 tools or do I need to write some additional tools? Currently, I am able to write/read the data to/from NV with the owner account and lock it for further reads but I need to be sure the value can be released from TPM just in case the platform is still trustworthy.

Thanks in advance
Fis

**Comment by atomsoftwarestudios** [comment]
a) problem solved - It was necessary to set SHA-1 algorithm in BIOS instead of SHA-2.

b) I have found the tpm2_nvdefine does not allow to set the algorithm to be used for access policy computation and sets 0 length for the digest field instead. I'll update it appropriately.

But still, I have no clue how the policy should be set. From the spec I see the name of the object (NV index) should be used for this purpose, but I have no clue where to use it. I'll continue with reading of the TCG spec.

**Comment by atomsoftwarestudios** [comment]
Ok, I think I got it. I will need to change the tpm2_nvdefine to allow usage of the policy digest. Additionally, I need to implement at least following tools in order to be possible to use the auth policy session for PCR authentization to be used when accessing the NV index:

tpm2_startauthsession
tpm2_policypcr
tpm2_getpolicydigest

**Comment by gwei3** [comment]
Good progress. Welcome to contribute!

**Comment by mikzaq** [comment]
@atomsoftwarestudios did you ever succeed in implementing those three tools?

**Comment by atomsoftwarestudios** [comment]
No. As I wrote in different issue I have discovered another implementation of TPM tools from IBM which we used. Google for IBM TPM2.0 tools.

**Comment by mikzaq** [comment]
I have the IBM TPM2.0 tools working, but I'm still having difficulty creating a pc policy and sealing to it. From the documentation it seems as though the steps are:
TPM2_StartAuthSession
TPM2_PolicyPCR (to create policy)
TPM2_GetRandom (to create key)
TPM2_Create
TPM2_Load

However, I keep getting an error that says my policy which I pass to TPM2_Create is the wrong size. Could you give me an example of the steps you went through?

**Comment by sagrawal77** [comment]
I used tpm2_listpcrs.cpp sample and modified it to create a PCR policy. 
Usage :  ../tpm2_buildpcrpolicy -L 0x04:1,16 -o pcr_policy.bin

I used the pcr_policy.bin in creating a RSA key which I used for signing. I also had to update the existing tpm2_sign.cpp to use PCR policysession instead of TPM_RS_PW to use the above RSA key.

I am attaching here the code for tpm2_buildpcrpolicy.cpp. Hopefully somebody will find it useful.


[tpm2_buildpcrpolicy.zip](https://github.com/01org/tpm2.0-tools/files/488099/tpm2_buildpcrpolicy.zip)
